cpp/src/arrow/ipc/writer.cc

```diff
@@ -1278,53 +1226,95 @@ Result<std::shared_ptr<Buffer>> SerializeRecordBatch(const RecordBatch& batch,
   ARROW_ASSIGN_OR_RAISE(auto buffer, mm->AllocateBuffer(size));
   ARROW_ASSIGN_OR_RAISE(auto writer, Buffer::GetWriter(buffer));
 
-  MemoryPool* pool;
+  IpcOptions options;
   // XXX Should we have a helper function for getting a MemoryPool
   // for any MemoryManager (not only CPU)?
   if (mm->is_cpu()) {
-    pool = checked_pointer_cast<CPUMemoryManager>(mm)->pool();
-  } else {
-    // Allocations will be ephemeral anyway
-    pool = default_memory_pool();
+    options.memory_pool = checked_pointer_cast<CPUMemoryManager>(mm)->pool();
   }
-  RETURN_NOT_OK(SerializeRecordBatch(batch, pool, writer.get()));
+  RETURN_NOT_OK(SerializeRecordBatch(batch, options, writer.get()));
   RETURN_NOT_OK(writer->Close());
   return buffer;
 }
 
-Status SerializeRecordBatch(const RecordBatch& batch, MemoryPool* pool,
+Status SerializeRecordBatch(const RecordBatch& batch, const IpcOptions& options,
                             std::shared_ptr<Buffer>* out) {
   int64_t size = 0;
   RETURN_NOT_OK(GetRecordBatchSize(batch, &size));
   std::shared_ptr<Buffer> buffer;
-  RETURN_NOT_OK(AllocateBuffer(pool, size, &buffer));
+  RETURN_NOT_OK(AllocateBuffer(options.memory_pool, size, &buffer));
 
   io::FixedSizeBufferWriter stream(buffer);
-  RETURN_NOT_OK(SerializeRecordBatch(batch, pool, &stream));
+  RETURN_NOT_OK(SerializeRecordBatch(batch, options, &stream));
   *out = buffer;
   return Status::OK();
 }
 
-Status SerializeRecordBatch(const RecordBatch& batch, MemoryPool* pool,
+Status SerializeRecordBatch(const RecordBatch& batch, const IpcOptions& options,
                             io::OutputStream* out) {
-  auto options = IpcOptions::Defaults();
   int32_t metadata_length = 0;
   int64_t body_length = 0;
-  return WriteRecordBatch(batch, 0, out, &metadata_length, &body_length, options, pool);
+  return WriteRecordBatch(batch, 0, out, &metadata_length, &body_length, options);
 }
 
 Status SerializeSchema(const Schema& schema, DictionaryMemo* dictionary_memo,
                        MemoryPool* pool, std::shared_ptr<Buffer>* out) {
   ARROW_ASSIGN_OR_RAISE(auto stream, io::BufferOutputStream::Create(1024, pool));
 
   auto options = IpcOptions::Defaults();
-  auto payload_writer = make_unique<PayloadStreamWriter>(options, stream.get());
-  RecordBatchPayloadWriter writer(std::move(payload_writer), schema, options,
-                                  dictionary_memo);
+  internal::IpcFormatWriter writer(
+      make_unique<internal::PayloadStreamWriter>(stream.get()), schema, options,
```

Comment:
```suggestion
      internal::make_unique<internal::PayloadStreamWriter>(stream.get()), schema, options,
```
Use of `make_unique` without `internal::` has caused issues on older MSVC before https://jira.apache.org/jira/browse/ARROW-5121 (TL;DR: it gets confused between our `internal::make_unique` and `std::make_unique` even if you have explicitly `using`'d only one of them)